src/sql/SqlResult.ts

```diff
@@ -0,0 +1,393 @@
+     */
+    private closed: boolean;
+
+    /**
+     * Row metadata of the result. Initially null.
+     */
+    private rowMetadata: SqlRowMetadata | null;
+
+    constructor(
+        private readonly sqlService: SqlServiceImpl,
+        private readonly serializationService: SerializationService,
+        private readonly connection: Connection,
+        private readonly queryId: SqlQueryId,
+        /** The page size used for pagination */
+        private readonly cursorBufferSize: number,
+        /** If true, SqlResult is an object iterable, otherwise SqlRow iterable */
+        private readonly returnRawResult: boolean,
+        private readonly clientUUID: UUID
+    ) {
+        this.closed = false;
+        this.last = false;
+        this.rowMetadata = null;
+        this.currentPage = null;
+        this.executeDeferred = deferredPromise<boolean>();
+        this.executeDeferred.promise.catch(() => {
+        });
+    }
+
+    /** This symbol is needed to be included to be an async iterable */
+    [Symbol.asyncIterator](): AsyncIterator<SqlRowType, SqlRowType, SqlRowType> {
+        const nextFn = this.next.bind(this);
+        return {
+            next: nextFn
+        }
+    }
+
+    /**
+     * Useful for mocking. (Constructor mocking is hard/impossible)
+     * @returns new result object.
+     */
+    static newResult(
+        sqlService: SqlServiceImpl,
+        serializationService: SerializationService,
+        connection: Connection,
+        queryId: SqlQueryId,
+        cursorBufferSize: number,
+        returnRawResult: boolean,
+        clientUUID: UUID
+    ) {
+        return new SqlResultImpl(sqlService, serializationService, connection, queryId, cursorBufferSize,
+            returnRawResult, clientUUID);
+    }
+
+    getUpdateCount(): Promise<Long> {
+        return this.executeDeferred.promise.then(() => {
+            return this.updateCount;
+        });
+    }
+
+    isRowSet(): Promise<boolean> {
+        return this.executeDeferred.promise.then(() => {
+            return this.updateCount === Long.fromInt(-1);
+        });
+    }
+
+    getRowMetadata(): Promise<SqlRowMetadata | null> {
+        return this.executeDeferred.promise.then(() => {
+            return this.rowMetadata;
+        });
+    }
+
+    close(): Promise<void> {
+        // Return the close promise if a close request is already started
+        if (this.closeDeferred?.promise) {
+            return this.closeDeferred.promise;
+        }
+
+        // If already closed, return a resolved promise
+        if (this.closed) {
+            return Promise.resolve();
+        }
+
+        this.closeDeferred = deferredPromise<void>();
+
+        const error = new HazelcastSqlException(this.clientUUID, SqlErrorCode.CANCELLED_BY_USER,
+            'Query was cancelled by user');
+        // Reject execute with user cancellation error.
+        this.onExecuteError(error);
```

Comment:
At this point, `this.closed` is false, so we will set some properties twice. We should set the closed flag to true above (maybe just below the if check)

Reply:
we are setting it below sqlService.close(), can you explain the setting properties twice case?

Reply:
Let's say they initialized once (when the server sends a response), then you call `close()`. On the point we call `this.onExecuteError(error);`, `this.closed` is false. So, in `onExecuteError`, we set the properties twice

Reply:
looking at the Java side, I think we should have a variable that holds if a response/error is received or not. we should call onExecuteError, only if a response is not received. in java side `state` is the variable that is checked.

Reply:
updateCount can be used but I don't think it is more understandable than introducing a new variable.

Reply:
I think we don't need it. Just set the `this.closed = true;` before(below the place you do the if check), not after, `onExecuteError` and it should be ok

Reply:
moving `this.closed = true` above will make `onExecuteError` no op for all cases.

Reply:
Oh, true

Reply:
I think an extra variable should be OK, it will be an internal thing. I think it is better than showing incorrect results to the user

Reply:
Sent the commit, you can check it

Reply:
Looks good